### PR TITLE
Make current test-proxy-master simple

### PR DESCRIPTION
**Pull request by hatoo**
@ohaddahan 
Hi, I think proxy-master can be more simple like this.
Please tell me if this architecture isn't enough for your plan or constrain.

I'll think based on that.

### CODE REVIEW

**Comment by ohaddahan** [comment]
Oh wow @hatoo this is much simpler and better.
Didn't know you can use a `unbounded_channel` with `copy_bidirectional`.

**Comment by ohaddahan** [comment]
I wonder, if we have multiple different `proxy-endpoint` and we want to decide to which one to send.
Let's say we have 2 `proxy-endpoint` and we want to send all requests to `https://example.com` to one, and all requests to `ifconfig.me` to the other.

**Comment by ohaddahan** [comment]
Few news.

1. Wow, awesome change by you. I didn't know sending the entire `TcpStream` inside a `channel` is possible. Added explicit type annotation for clarity.
2. Is it possible to read headers inside `proxy-master`? Of the incoming/returning requests?
3. Is it possible to hold multiple input connections from more than one `proxy-endpoint` and direct to them by some logic? For simplicity something like https://github.com/block-mesh/block-mesh-monorepo/pull/56#issuecomment-2016530142

**Comment by hatoo** [comment]
> 2. Is it possible to read headers inside proxy-master? Of the incoming/returning requests? Something like a middleware so we can track/edit usage/headers.

I've chosen to use `httparse` to read the client header because `hyper` server has some difficulties of upgrading connection in this situation

**Comment by hatoo** [comment]
I've added 2. 3. 4. now

File: libs/test-proxy-master/src/main.rs [hunk 1]
```diff
@@ -1,68 +1,121 @@
...
+                        // Write the rest of the buffer
+                        proxy.write_all(&buf[len..]).await?;
+                        tokio::io::copy_bidirectional(&mut stream, &mut proxy).await?;
+                        break;
+                    }
+                }
+
+                Ok::<(), std::io::Error>(())
+            });
+        }
```
Comment (hatoo) [inline libs/test-proxy-master/src/main.rs:78]: I don't like this part.
Because it uses a kinda manual parsing here.
But this looks unavoidable for now

**Comment by ohaddahan** [comment]
```rust
    let pool1 = pool.clone();
    let task_1 = tokio::task::spawn(async move {
        while let Ok((stream, _addr)) = proxy_listener.accept().await {
            let pool = pool1.clone();
            if let Err(err) = server::conn::http1::Builder::new()
                .preserve_header_case(true)
                .title_case_headers(true)
                .serve_connection(
                    TokioIo::new(stream),
                    service_fn(move |req| proxy_endpoint(pool.clone(), req)),
                )
                .with_upgrades()
                .await
            {
                println!("Failed to serve connection: {:?}", err);
            }
        }
    });
```

I think we can add a `tokio::spawn` here inside each iteration so it won't block and be able to handle multiple concurrently, right?
I'm adding it.

File: libs/test-proxy-master/src/main.rs (hunk 1 above)
Comment (ohaddahan) [reply libs/test-proxy-master/src/main.rs]: Yeah I agree, I'm trying to play with it and see if something else can be used.

**Comment by ohaddahan** [comment]
Did a bit of refactoring of splitting the code to multiple files.

**Comment by ohaddahan** [comment]
I noticed two things:

1. If you send a request `curl 127.0.0.1:4000` to the `proxy-master` , the `proxy-endpoint` will still receive it since at the moment we don't do any processing in `listen_for_clients_connecting`.

2. I think in `listen_for_clients_connecting` , if we do a symmetrical method to `listen_for_proxies_connecting` , which runs through `serve_connection` we'll be able to:
  2.1. Edit the request headers easily.
  2.2. Avoid issue (1).


I tried doing it but couldn't get it to work yet, so I reverted the code for now.

**Comment by hatoo** [comment]
> ```rust
>     let pool1 = pool.clone();
>     let task_1 = tokio::task::spawn(async move {
>         while let Ok((stream, _addr)) = proxy_listener.accept().await {
>             let pool = pool1.clone();
>             if let Err(err) = server::conn::http1::Builder::new()
>                 .preserve_header_case(true)
>                 .title_case_headers(true)
>                 .serve_connection(
>                     TokioIo::new(stream),
>                     service_fn(move |req| proxy_endpoint(pool.clone(), req)),
>                 )
>                 .with_upgrades()
>                 .await
>             {
>                 println!("Failed to serve connection: {:?}", err);
>             }
>         }
>     });
> ```
> 
> I think we can add a `tokio::spawn` here inside each iteration so it won't block and be able to handle multiple concurrently, right? I'm adding it.

Yes, you are right 👍

**Pull request merged**